flow/diff_context.cc

```diff
@@ -128,23 +145,32 @@ Damage DiffContext::ComputeDamage(const SkIRect& accumulated_buffer_damage,
   return res;
 }
 
+SkRect DiffContext::MapRect(const SkRect& rect) {
+  SkRect mapped_rect;
+  if (!clip_tracker_.mapRect(rect, &mapped_rect)) {
+    // TODO(knopp): What to do here? We could possibly force repainting
+    // the whole frame?
+  }
+  return mapped_rect;
+}
+
 bool DiffContext::PushCullRect(const SkRect& clip) {
-  SkRect cull_rect = state_.transform.mapRect(clip);
-  return state_.cull_rect.intersect(cull_rect);
+  clip_tracker_.clipRect(clip, SkClipOp::kIntersect, false);
+  return !clip_tracker_.device_cull_rect().isEmpty();
 }
 
-SkRect DiffContext::GetCullRect() const {
-  SkMatrix inverse_transform;
-  // Perspective projections don't produce rectangles that are useful for
-  // culling for some reason.
-  if (!state_.transform.hasPerspective() &&
-      state_.transform.invert(&inverse_transform)) {
-    return inverse_transform.mapRect(state_.cull_rect);
+SkMatrix DiffContext::GetTransform3x3() const {
+  if (clip_tracker_.using_4x4_matrix()) {
```

Comment:
Just `return clip_tracker_.matrix_3x3()`. It does the down-convert as necessary.